File: pkg/api/src/main/java/ode/model/PermissionsI.java
package ode.model;

// Copyright (c) 2018 Bhojpur Consulting Private Limited, India. All rights reserved.

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import static ode.model.internal.Permissions.ANNOTATERESTRICTION;
import static ode.model.internal.Permissions.CHGRPRESTRICTION;
import static ode.model.internal.Permissions.CHOWNRESTRICTION;
import static ode.model.internal.Permissions.DELETERESTRICTION;
import static ode.model.internal.Permissions.EDITRESTRICTION;
import static ode.model.internal.Permissions.LINKRESTRICTION;
import java.util.Arrays;
import ode.util.Utils;
import Ice.Current;
import Ice.Object;

/**
 * Bhojpur ODE server wrapper around the {@link ode.model.internal.Permissions} class.
 * Currently, the internal representation is made public. (see the ZeroC thread
 * link below), but should not be used by clients.
 *
 * @see <a href="http://www.zeroc.com/forums/showthread.php?t=3084">ZeroC Thread
 *      3084</a>
 *
 */
public class PermissionsI extends Permissions implements ode.model.ModelBased {

    private static final long serialVersionUID = 89928049580980928L;

    public final static Ice.ObjectFactory Factory = new Ice.ObjectFactory() {

        public Object create(String arg0) {
            return new PermissionsI();
        }

        public void destroy() {
            // no-op
        }

    };

    /**
     * Whether or not this object is immutable. Currently this should only
     * be set to true after marshalling and unmarshalling via Ice.
     */
    private boolean __immutable = false;

    /**
     * Called as Ice converts from a binary stream to a PermissionsI object.
     * Here we set {@link #__immutable} to true so that clients consuming this
     * object cannot alter them.
     */
    @Override
    public void ice_postUnmarshal() {
        super.ice_postUnmarshal();
        __immutable = true;
    }

    public PermissionsI() {
        Long l = (Long) Utils
                .internalForm(ode.model.internal.Permissions.EMPTY);
        if (l == null) {
            throw new IllegalStateException("Permissions.DEFAULT is null");
        }
        this.perm1 = l.longValue();
    }

    public PermissionsI(String representation) {
        Long l = (Long) Utils.internalForm(ode.model.internal.Permissions
                .parseString(representation));
        if (l == null) {
            throw new IllegalStateException(representation + " produced null");
        }
        this.perm1 = l.longValue();
    }

    @Override
    public boolean isRestricted(String restriction, Current __current) {
        if (extendedRestrictions != null) {
            return extendedRestrictions.contains(restriction);
        }
        return false;
    }

    public boolean isDisallow(final int restriction, final Ice.Current c) {
        return ode.model.internal.Permissions
            .isDisallow(restrictions, restriction);
    }

    public boolean canAnnotate(final Ice.Current c) {
        return !isDisallow(ANNOTATERESTRICTION, c);
    }

    @Override
    public boolean canChgrp(final Ice.Current c) {
        return !isDisallow(CHGRPRESTRICTION, c);
    }

    @Override
    public boolean canChown(final Ice.Current c) {
        return !isDisallow(CHOWNRESTRICTION, c);
    }

    public boolean canDelete(final Ice.Current c) {
        return !isDisallow(DELETERESTRICTION, c);
    }

    public boolean canEdit(final Ice.Current c) {
        return !isDisallow(EDITRESTRICTION, c);
    }

    public boolean canLink(final Ice.Current c) {
        return !isDisallow(LINKRESTRICTION, c);
    }

    public PermissionsI(PermissionsI perms) {
        this.perm1 = perms.getPerm1();
    }

    public PermissionsI(ode.model.internal.Permissions sourceP) {
        setPerm1((Long) ode.util.Utils.internalForm(sourceP));
        this.restrictions = sourceP.getRestrictions();
        String[] extRestr = sourceP.copyExtendedRestrictions();
        this.extendedRestrictions = extRestr == null ? null :
                Arrays.<String>asList(extRestr);
    }

    public long getPerm1(Ice.Current current) {
        return this.perm1;
    }

    public void setPerm1(long perm1, Ice.Current current) {
        throwIfImmutable();
        this.perm1 = perm1;
    }

    public void setPerm1(Long perm1) {
        throwIfImmutable();
        this.perm1 = perm1 == null ? 0 : perm1.longValue();

    }

    public void copyObject(ode.util.Filterable model,
            ode.util.ModelMapper _mapper) {
        throw new UnsupportedOperationException();
    }

    public ode.util.Filterable fillObject(ode.util.ReverseModelMapper _mapper) {
        throw new UnsupportedOperationException();
    }

    public void unload(Ice.Current c) {
        this.setPerm1(null);
    }

    // shift 8; mask 4
    public boolean isUserRead(Ice.Current c) {
        return granted(4, 8);
    }

    public void setUserRead(boolean value, Ice.Current c) {
        set(4, 8, value);
    }

    // shift 8; mask 2
    public boolean isUserWrite(Ice.Current c) {
        return granted(2, 8);
    }

    public void setUserWrite(boolean value, Ice.Current c) {
        set(2, 8, value);
    }

    // shift 8; mask 1
    public boolean isUserAnnotate(Ice.Current c) {
        return granted(1, 8);
    }

    public void setUserAnnotate(boolean value, Ice.Current c) {
        set(1, 8, value);
    }

    // shift 4; mask 4
    public boolean isGroupRead(Ice.Current c) {
        return granted(4, 4);
    }

    public void setGroupRead(boolean value, Ice.Current c) {
        set(4, 4, value);
    }

    // shift 4; mask 2
    public boolean isGroupWrite(Ice.Current c) {
        return granted(2, 4);
    }

    public void setGroupWrite(boolean value, Ice.Current c) {
        set(2, 4, value);
    }

    // shift 4; mask 1
    public boolean isGroupAnnotate(Ice.Current c) {
        return granted(1, 4);
    }

    public void setGroupAnnotate(boolean value, Ice.Current c) {
        set(1, 4, value);
    }

    // shift 0; mask 4
    public boolean isWorldRead(Ice.Current c) {
        return granted(4, 0);
    }

    public void setWorldRead(boolean value, Ice.Current c) {
        set(4, 0, value);
    }

    // shift 0; mask 2
    public boolean isWorldWrite(Ice.Current c) {
        return granted(2, 0);
    }

    public void setWorldWrite(boolean value, Ice.Current c) {
        set(2, 0, value);
    }

    // shift 0; mask 1
    public boolean isWorldAnnotate(Ice.Current c) {
        return granted(1, 0);
    }

    public void setWorldAnnotate(boolean value, Ice.Current c) {
        set(1, 0, value);
    }

    protected boolean granted(int mask, int shift) {
        return (perm1 & (mask << shift)) == (mask << shift);
    }

    protected void set(int mask, int shift, boolean on) {
        throwIfImmutable();
        if (on) {
            perm1 = perm1 | (0L | (mask << shift));
        } else {
            perm1 = perm1 & (-1L ^ (mask << shift));
        }
    }

    private void throwIfImmutable() {
        if (__immutable) {
            throw new ode.ClientError("ImmutablePermissions:"+toString());
        }
    }

    // ~ Overrides
    // =========================================================================

    /**
     * produces a String representation of the {@link PermissionsI} similar to
     * those on a Unix filesystem. Unset bits are represented by a dash, while
     * other bits are represented by a symbolic value in the correct bit
     * position. For example, a Permissions with all rights
     * granted to all but WORLD roles would look like: rwrw--
     */
    @Override
    public String toString() {
        return Utils.toPermissions(perm1).toString();
    }

}

File: pkg/components/tools/OdeCpp/src/ode/model/PermissionsI.cpp
// Copyright (c) 2018 Bhojpur Consulting Private Limited, India. All rights reserved.

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include <ode/ClientErrors.h>
#include <ode/model/PermissionsI.h>
#include <ode/ClientErrors.h>
#include <ode/Constants.h>

::Ice::Object* IceInternal::upCast(::ode::model::PermissionsI* p) { return p; }

namespace ode {

    namespace model {

        PermissionsI::~PermissionsI() {}

        PermissionsI::PermissionsI(const std::string& perms) : Permissions() {
            __immutable = false;
            restrictions = ode::api::BoolArray();
            if (perms.empty()) {
                perm1 = -1L;
            } else {
                perm1 = 0L;
                // Due to lack of regex libraries
                // this implementation is more naive
                // than in the other SDKs
                std::string rest = perms;
                if (rest.size() == 7) {
                    // ignore the locked character
                    rest = rest.substr(1);
                }
                if (rest.size() != 6) {
                    std::string msg = "Bad permissions:";
                    msg += perms;
                    throw ode::ClientError(__FILE__, __LINE__, msg.c_str());
                }
                if (rest[0] == 'r' || rest[0] == 'R') {
                    setUserRead(true);
                }
                if (rest[1] == 'w' || rest[1] == 'W') {
                    setUserWrite(true);
                    setUserAnnotate(true);
                } else if (rest[1] == 'a' || rest[1] == 'A') {
                    setUserAnnotate(true);
                }
                if (rest[2] == 'r' || rest[2] == 'R') {
                    setGroupRead(true);
                }
                if (rest[3] == 'w' || rest[3] == 'W') {
                    setGroupWrite(true);
                    setGroupAnnotate(true);
                } else if (rest[3] == 'a' || rest[3] == 'A') {
                    setGroupAnnotate(true);
                }
                if (rest[4] == 'r' || rest[4] == 'R') {
                    setWorldRead(true);
                }
                if (rest[5] == 'w' || rest[5] == 'W') {
                    setWorldWrite(true);
                    setWorldAnnotate(true);
                } else if (rest[5] == 'a' || rest[5] == 'A') {
                    setWorldAnnotate(true);
                }
            }
        }

        void PermissionsI::ice_postUnmarshal() {
            __immutable = true;
        }

        void PermissionsI::throwIfImmutable() {
            if (__immutable) {
                throw ode::ClientError(__FILE__,__LINE__,"ImmutablePermissions");
            }
        }

        void PermissionsI::unload(const Ice::Current& /* current */) {
            // no-op
        }

        bool PermissionsI::isRestricted(const std::string& restriction, const Ice::Current& /* current */) {
            const ode::api::StringSet::const_iterator beg = extendedRestrictions.begin();
            const ode::api::StringSet::const_iterator end = extendedRestrictions.end();
            return std::find(beg, end, restriction) != end;
        }

        bool PermissionsI::isDisallow(::Ice::Int restriction, const Ice::Current& /* current */) {
            // restriction should be unsigned but that is not what's generated by Ice.
            // Instead for any negative value we return false, since there can be no
            // element in restrictions which is true.
            if (restriction < 0) {
                return false;
            }

            // and if it is positive, then it's safe to cast to unsigned.
            if (restrictions.size() > static_cast<unsigned int>(restriction)) {
                return restrictions[restriction];
            }
            return false;
        }

        bool PermissionsI::canAnnotate(const Ice::Current& /* current */) {
            return ! isDisallow(ode::constants::permissions::ANNOTATERESTRICTION);
        }

        bool PermissionsI::canDelete(const Ice::Current& /* current */) {
            return ! isDisallow(ode::constants::permissions::DELETERESTRICTION);
        }

        bool PermissionsI::canEdit(const Ice::Current& /* current */) {
            return ! isDisallow(ode::constants::permissions::EDITRESTRICTION);
        }

        bool PermissionsI::canLink(const Ice::Current& /* current */) {
            return ! isDisallow(ode::constants::permissions::LINKRESTRICTION);
        }

        bool PermissionsI::canChgrp(const Ice::Current& /* current */) {
            return ! isDisallow(ode::constants::permissions::CHGRPRESTRICTION);
        }

        bool PermissionsI::canChown(const Ice::Current& /* current */) {
            return ! isDisallow(ode::constants::permissions::CHOWNRESTRICTION);
        }

        Ice::Long PermissionsI::getPerm1(const Ice::Current& /* current */) {
            return perm1;
        }

        void PermissionsI::setPerm1(Ice::Long _perm1, const Ice::Current& /* current */) {
            throwIfImmutable();
            perm1 =  _perm1;
        }

        // shift 8; mask 4
        bool PermissionsI::isUserRead(const Ice::Current& /* current */) {
            return granted(4,8);
        }
        void PermissionsI::setUserRead(bool value, const Ice::Current& /* current */) {
            set(4,8, value);
        }

        // shift 8; mask 2
        bool PermissionsI::isUserWrite(const Ice::Current& /* current */) {
            return granted(2,8);
        }
        void PermissionsI::setUserWrite(bool value, const Ice::Current& /* current */) {
            set(2,8, value);
        }

        // shift 8; mask 1
        bool PermissionsI::isUserAnnotate(const Ice::Current& /* current */) {
            return granted(1,8);
        }
        void PermissionsI::setUserAnnotate(bool value, const Ice::Current& /* current */) {
            set(1,8, value);
        }

        // shift 4; mask 4
        bool PermissionsI::isGroupRead(const Ice::Current& /* current */) {
            return granted(4,4);
        }
        void PermissionsI::setGroupRead(bool value, const Ice::Current& /* current */) {
            set(4,4, value);
        }

        // shift 4; mask 2
        bool PermissionsI::isGroupWrite(const Ice::Current& /* current */) {
            return granted(2,4);
        }
        void PermissionsI::setGroupWrite(bool value, const Ice::Current& /* current */) {
            set(2,4, value);
        }

        // shift 4; mask 1
        bool PermissionsI::isGroupAnnotate(const Ice::Current& /* current */) {
            return granted(1,4);
        }
        void PermissionsI::setGroupAnnotate(bool value, const Ice::Current& /* current */) {
            set(1,4, value);
        }

        // shift 0; mask 4
        bool PermissionsI::isWorldRead(const Ice::Current& /* current */) {
            return granted(4,0);
        }
        void PermissionsI::setWorldRead(bool value, const Ice::Current& /* current */) {
            set(4,0, value);
        }

        // shift 0; mask 2
        bool PermissionsI::isWorldWrite(const Ice::Current& /* current */) {
            return granted(2,0);
        }
        void PermissionsI::setWorldWrite(bool value, const Ice::Current& /* current */) {
            set(2,0, value);
        }

        // shift 0; mask 1
        bool PermissionsI::isWorldAnnotate(const Ice::Current& /* current */) {
            return granted(1,0);
        }
        void PermissionsI::setWorldAnnotate(bool value, const Ice::Current& /* current */) {
            set(1,0, value);
        }

        bool PermissionsI::granted(int mask, int shift) {
            return (perm1 & (mask<<shift) ) == (mask<<shift);
        }

        void PermissionsI::set(int mask, int shift, bool on) {
            throwIfImmutable();
            if (on) {
                perm1 = perm1 | ( 0L  | (mask<<shift) );
            } else {
                perm1 = perm1 & ( -1L ^ (mask<<shift) );
            }
        }

    }
} //End ode::model